change(state): Refactor format upgrades into trait

arya2 opened the pull request:
## Motivation

This PR organizes Zebra's disk format upgrades into trait implementations to make them more maintainable as more upgrades are gradually added.

Closes #7932

## Solution

- Add a new `DiskFormatUpgrade` trait
- Implement the trait for each existing disk format upgrade (except for the indexer upgrade which is handled outside the usual db format upgrade process)

### Tests

This change should be covered by existing tests for applying db format upgrades.

### Follow-up Work

The bundle of db format changes needed to match zcashd's RPC outputs.

### PR Author's Checklist

<!-- If you are the author of the PR, check the boxes below before making the PR
ready for review. -->

- [x] The PR name will make sense to users.
- [ ] The PR provides a CHANGELOG summary.
- [x] The solution is tested.
- [x] The documentation is up to date.
- [x] The PR has a priority label.

### PR Reviewer's Checklist

<!-- If you are a reviewer of the PR, check the boxes below before approving it. -->

- [ ] The PR Author's checklist is complete.
- [ ] The PR resolves the issue.

arya2 committed:
Adds a new trait for disk format upgrades, implements in on a new struct, `PruneTrees`, and moves the logic for tree deduplication to the trait impl

arya2 committed:
refactors add subtrees format upgrade to use new trait

arya2 committed:
refactors fix tree keys, cache genesis roots, and value balance upgrades to use new trait

arya2 commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -20,21 +20,78 @@ use zebra_chain::{
+    fn description(&self) -> &'static str;
+
+    /// Runs disk format upgrade.
+    fn run(
```
`run_migration()` may be a better name for this method.

conradoplg replied:
Both are fine I think.

arya2 commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -20,21 +20,78 @@ use zebra_chain::{
+    ///
+    /// # Panics
+    ///
+    /// If the state has not been upgraded to this format correctly.
```
```suggestion
```

conradoplg replied:
+1

conradoplg replied:
Also would be good to explain the return value, e.g.

```
The outer Result indicates whether the validation was cancelled (due to e.g. node shutdown). The inner Result indicates whether the validation itself failed or not.
```

conradoplg commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -20,21 +20,78 @@ use zebra_chain::{
+}
+
+fn format_upgrades() -> Vec<Box<dyn DiskFormatUpgrade>> {
+    // Note: Disk format upgrades must be run in order.
```
I think it would be good to check if they upgrades are really in order and panic if they're not

```suggestion
    // Note: Disk format upgrades must be run in order of database version.
```

arya2 replied:
Applied the doc change and added a test to check that they're in order in https://github.com/ZcashFoundation/zebra/pull/9263/commits/55b36028eb45e0349f56f7ebb24088f1f6c53e4e.

conradoplg commented on zebra-state/src/service/finalized_state/disk_format/upgrade/no_migration.rs:
```diff
@@ -0,0 +1,48 @@
+
+use super::{CancelFormatChange, DiskFormatUpgrade};
+
+/// Implements [`DiskFormatUpgrade`] for pruning duplicate Sapling and Orchard note commitment trees from database
```
Wrong docs?

arya2 replied:
Yep, thank you, fixed in https://github.com/ZcashFoundation/zebra/pull/9263/commits/55b36028eb45e0349f56f7ebb24088f1f6c53e4e.

conradoplg commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -474,132 +531,39 @@ impl DbFormatChange {
+                upgrade
+                    .validate(db, cancel_receiver)?
+                    .expect("failed to validate db format");
+                continue;
```
I'm a bit confused about this. The original code does not seem to do validate if the disk version >= upgrade version. Is this required or just good to have? What is the performance impact?

arya2 replied:
This change was wrong, I thought some of the upgrades were being validated here even when they weren't being applied, but none are.

> What is the performance impact?

It shouldn't have a significant performance impact, it won't block anything. These will still always run at startup anyway but should run after any required format upgrades have been applied.

conradoplg replied:
I'm assuming you're still working on this then?

arya2 replied:
Removed in https://github.com/ZcashFoundation/zebra/pull/9263/commits/55b36028eb45e0349f56f7ebb24088f1f6c53e4e (by accepting a `min_version` in `format_upgrades()` and filtering out older upgrades altogether).

Also removed the old code for marking the database as upgraded to the value balance format (26.0.0) and restored the log message for that upgrade (and future no_migration upgrades)

conradoplg reviewed (commented):
This looks good overall, but I'm confused about some parts. (Sorry, not super familiar with the original code)

arya2 committed:
Applies suggestions from code review:
- Avoids duplicate validation of format upgrades at startup when db is already upgraded,
- Minor refactors
- Doc fixes and cleanups

conradoplg commented on zebra-state/src/service/finalized_state/disk_format/upgrade/prune_trees.rs:
```diff
@@ -0,0 +1,149 @@
+    ///
+    /// # Panics
+    ///
+    /// If a duplicate tree is found.
```
```suggestion
```

I don' think this panics? The caller does that

conradoplg reviewed (dismissed):
Looks great, thanks! I left just a couple of documentation suggestions.

Sorry for taking this long - I wanted to learn about the overall database upgrade process since I'm also going to work with it

arya2 committed:
Applies suggestions from code review

oxarbitrage commented:
@Mergifyio update

mergify[bot] committed:
Merge branch 'main' into db-format-upgrade-trait
